Shard blob, headers and slashings (#2302)
ethereum/eth2.0-specs: specs/sharding/p2p-interface.md

## New containers

### ShardBlobBody

```python
class ShardBlobBody(Container):
    # The actual data commitment
    commitment: DataCommitment
    # Proof that the degree < commitment.length
    degree_proof: BLSCommitment
    # The actual data. Should match the commitment and degree proof.
    data: List[BLSPoint, POINTS_PER_SAMPLE * MAX_SAMPLES_PER_BLOCK]
```

### ShardBlob

```python
class ShardBlob(Container):
    # Slot and shard that this header is intended for
    slot: Slot
    shard: Shard
    body: ShardBlobBody
    # Proposer of the shard-blob
    proposer_index: ValidatorIndex
```

> The user MUST always verify the `body` matches the commitments in the `commitment` and `degree_proof`.

Author: This is worded poorly, inconsistent after latest type changes. Meant to require consistency between commitments and data in `body`

Author: Fixed it, and moved the note to the body definition.

## New containers

### ShardBlobBody

```python

> class ShardBlobBody(Container):

Reviewer: I lean toward moving these into the consensus spec, but I see the argument for not

Reviewer: ehh. maybe not

    # The actual data commitment
    commitment: DataCommitment
    # Proof that the degree < commitment.length
    degree_proof: BLSCommitment
    # The actual data. Should match the commitment and degree proof.
    data: List[BLSPoint, POINTS_PER_SAMPLE * MAX_SAMPLES_PER_BLOCK]
```

> The user MUST always verify the commitments in the `body` are valid for the `data` in the `body`.

Reviewer 2: Can we clarify what the commitments are? Is it `body.commitment`? if yes, maybe this is more clear:

```suggestion
The user MUST always verify the `commitment` in the `body` is valid for the `data` in the `body`.
```

Author: "commitments" here refers to both KZG commitments, the only two in that type. The first one checks the contents, the second one checks the length. I will see if we can add more comments, or explain the commitments elsewhere in the spec.

### ShardBlob

```python
class ShardBlob(Container):

>     # Slot and shard that this header is intended for

Reviewer: Accidental find and replace? This is a `Blob` not a header

| `shard_header`                   | `SignedShardHeader`       |
| `shard_proposer_slashing`        | `ShardProposerSlashing`   |

The [DAS network specification](./das-p2p.md) defines additional topics.

#### Shard blobs: `shard_blob_{shard}`

Shard block data, in the form of a `SignedShardBlob` is published to the `shard_blob_{shard}` subnets.

The following validations MUST pass before forwarding the `signed_blob` (with inner `message` as `blob`) on the horizontal subnet or creating samples for it.
- _[REJECT]_ `blob.shard` MUST match the topic `{shard}` parameter. (And thus within valid shard index range)
- _[IGNORE]_ The `blob` is not from a future slot (with a `MAXIMUM_GOSSIP_CLOCK_DISPARITY` allowance) --
  i.e. validate that `blob.slot <= current_slot`
  (a client MAY queue future blobs for processing at the appropriate slot).
- _[IGNORE]_ The `blob` is new enough to be still be processed --
  i.e. validate that `compute_epoch_at_slot(blob.slot) >= get_previous_epoch(state)`
- _[IGNORE]_ The blob is the first blob with valid signature received for the `(header.proposer_index, header.slot, header.shard)` combination.

> - _[REJECT]_ As already limited by the SSZ list-limit, it is important the blob is well-formatted and not too large.

Author: We could verify if the commitment and degree proof are valid for the contained data, but it wouldn't really matter much if it's not: double proposals are slashable, and the proposal is just wasted. So it could be an application-only thing, and skipped during network validation. This verification is faster than computing the commitment from the data, so it may be ok to still verify it anyway?

- _[REJECT]_ `blob.shard` MUST match the topic `{shard}` parameter. (And thus within valid shard index range)
- _[IGNORE]_ The `blob` is not from a future slot (with a `MAXIMUM_GOSSIP_CLOCK_DISPARITY` allowance) --
  i.e. validate that `blob.slot <= current_slot`
  (a client MAY queue future blobs for processing at the appropriate slot).
- _[IGNORE]_ The `blob` is new enough to be still be processed --
  i.e. validate that `compute_epoch_at_slot(blob.slot) >= get_previous_epoch(state)`
- _[IGNORE]_ The blob is the first blob with valid signature received for the `(blob.proposer_index, blob.slot, blob.shard)` combination.
- _[REJECT]_ As already limited by the SSZ list-limit, it is important the blob is well-formatted and not too large.
- _[REJECT]_ The `blob.data` MUST NOT contain any point `p >= MODULUS`. Although it is a `uint256`, not the full 256 bit range is valid.
- _[REJECT]_ The proposer signature, `signed_blob.signature`, is valid with respect to the `proposer_index` pubkey.
- _[REJECT]_ The blob is proposed by the expected `proposer_index` for the blob's slot

>   in the context of the current shuffling (defined by `parent_root`/`slot`).

Reviewer: This might be an issue. The current ShardBlob does not have a beacon chain reference (I think what you are referring to as `parent_root`) in the definition and so there is no expected shuffling to be verified against.

I believe we *must* have this in the shardblob payload to properly do verifications

If we add it, we might also add `shard_blob_` gossip validation that the `parent_root` or `beacon_root` reference has been processed locally (with the may IGNORE on that condition instead of this condition here that has it implicit

Author: It's based on the beacon-block proposer index, forgot about the `parent_root` context. That's a problem indeed, it needs some context for any forking. If we add such context though, shard blobs cannot be replayed anymore on a chain that prunes out some earlier shard blob header inclusions (if we enforce the `parent_root` to be correct on-chain). This may be a good thing though?

Author: It references the new `beacon_block_root` field now.

  If the `proposer_index` cannot immediately be verified against the expected shuffling,
  the block MAY be queued for later processing while proposers for the blob's branch are calculated --
  in such a case _do not_ `REJECT`, instead `IGNORE` this message.

> #### Shard header: `shard_header`

Reviewer: Consider putting "global topics" under it's own sub-header like in altair and beacon specs
https://github.com/ethereum/eth2.0-specs/blob/dev/specs/altair/p2p-interface.md#global-topics

Author: The nesting gets too deep for my liking, the topics themselves don't get a unique element-id link anymore that way.

Shard header data, in the form of a `SignedShardBlobHeader` is published to the global `shard_header` subnet.

The following validations MUST pass before forwarding the `signed_shard_header` (with inner `message` as `header`) on the network.
- _[IGNORE]_ The `header` is not from a future slot (with a `MAXIMUM_GOSSIP_CLOCK_DISPARITY` allowance) --
  i.e. validate that `header.slot <= current_slot`
  (a client MAY queue future headers for processing at the appropriate slot).
- _[IGNORE]_ The `header` is new enough to be still be processed --
  i.e. validate that `compute_epoch_at_slot(header.slot) >= get_previous_epoch(state)`

> - _[IGNORE]_ The header is the first header with valid signature received for the `(header.proposer_index, header.slot, header.shard)` combination.

Reviewer: TODO: consider adding a `beacon_root` reference processed check

Author: Possible solution: https://github.com/ethereum/eth2.0-specs/pull/2312

Author: Merged that draft into this PR, there's a `beacon_block_root` field now. The key stays `proposer_index, slot, shard`, since multiple proposals of the same proposer in the same slot would be slashable, there must only be one.
